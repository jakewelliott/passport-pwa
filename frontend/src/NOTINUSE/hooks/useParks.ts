import { useQuery } from '@tanstack/react-query';
import type { Park } from './types';
import { api, type APIError } from './api';

export const parkKeys = {
  all: ['parks'] as const,
  lists: () => [...parkKeys.all, 'list'] as const,
  detail: (id: number) => [...parkKeys.all, 'detail', id] as const,
};

export const useParks = () => {
  return useQuery<Park[], APIError>({
    queryKey: parkKeys.lists(),
    queryFn: () => api.getParks(),
  });
};

export const usePark = (id: number) => {
  return useQuery<Park, APIError>({
    queryKey: parkKeys.detail(id),
    queryFn: () => api.getPark(id),
    enabled: !!id,
  });
};
