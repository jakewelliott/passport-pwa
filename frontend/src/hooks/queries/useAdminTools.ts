import {
    API_AUTH_URL,
    API_BUCKET_LIST_URL,
    API_PARKS_URL,
    API_TRAILS_URL,
    fetchDelete,
    fetchPost,
    fetchPut,
} from '@/lib/fetch';
import type { BucketListItem, Park, Trail, UserProfile } from '@/types';
import { useMutation } from '@tanstack/react-query';
import type { Cruddable } from '../useCRUD';

// ADAM: normally i think one file, one purpose, but this is a collection of admin tools
// so i think it's less ugly to have all in one file

// parks
export const useCreatePark = () => {
    return useMutation({
        mutationKey: ['admin', 'parks', 'create'],
        mutationFn: (park: Park) => fetchPost(API_PARKS_URL, park),
    });
};

export const useUpdatePark = () => {
    return useMutation({
        mutationKey: ['admin', 'parks', 'update'],
        mutationFn: (park: Park) => fetchPut(`${API_PARKS_URL}/${park.id}`, park),
    });
};

export const useDeletePark = () => {
    return useMutation({
        mutationKey: ['admin', 'parks', 'delete'],
        mutationFn: (park: Park) => fetchDelete(`${API_PARKS_URL}/${park.id}`),
    });
};

export const useParkCRUD: Cruddable<Park> = {
    createHook: useCreatePark,
    updateHook: useUpdatePark,
    deleteHook: useDeletePark,
};

// bucket list
export const useCreateBucketListItem = () => {
    return useMutation({
        mutationKey: ['admin', 'bucketList', 'create'],
        mutationFn: (bucketListItem: BucketListItem) => fetchPost(API_BUCKET_LIST_URL, bucketListItem),
    });
};
export const useUpdateBucketListItem = () => {
    return useMutation({
        mutationKey: ['admin', 'bucketList', 'update'],
        mutationFn: (bucketListItem: BucketListItem) =>
            fetchPut(`${API_BUCKET_LIST_URL}/${bucketListItem.id}`, bucketListItem),
    });
};

export const useDeleteBucketListItem = () => {
    return useMutation({
        mutationKey: ['admin', 'bucketList', 'delete'],
        mutationFn: (bucketListItem: BucketListItem) => fetchDelete(`${API_BUCKET_LIST_URL}/${bucketListItem.id}`),
    });
};

export const useBucketListCRUD: Cruddable<BucketListItem> = {
    createHook: useCreateBucketListItem,
    updateHook: useUpdateBucketListItem,
    deleteHook: useDeleteBucketListItem,
};

// trails
export const useCreateTrail = () => {
    return useMutation({
        mutationKey: ['admin', 'trails', 'create'],
        mutationFn: (trail: Trail) => fetchPost(API_TRAILS_URL, trail),
    });
};

export const useUpdateTrail = () => {
    return useMutation({
        mutationKey: ['admin', 'trails', 'update'],
        mutationFn: (trail: Trail) => fetchPut(`${API_TRAILS_URL}/${trail.id}`, trail),
    });
};

export const useDeleteTrail = () => {
    return useMutation({
        mutationKey: ['admin', 'trails', 'delete'],
        mutationFn: (trail: Trail) => fetchDelete(`${API_TRAILS_URL}/${trail.id}`),
    });
};

export const useTrailCRUD: Cruddable<Trail> = {
    createHook: useCreateTrail,
    updateHook: useUpdateTrail,
    deleteHook: useDeleteTrail,
};

// users

export const useDontCallMe = () => {
    // ADAM: this is a no-op for type safety
    // i normally would never do this but here we are
    return useMutation({
        mutationKey: ['admin', 'users', 'update'],
        mutationFn: (user: UserProfile) => Promise.reject(new Error('NEVER EVER EVER CALL THIS FUNCTION')),
    });
};

export const useUpdateUserRole = () => {
    return useMutation({
        mutationKey: ['admin', 'users', 'updateRole'],
        mutationFn: (user: UserProfile) => fetchPut(`${API_AUTH_URL}/${user.id}/role`, user),
    });
};

// ADAM: yeah and ill figure this one out later
export const useUpdateUserPassword = () => {
    return useMutation({
        mutationKey: ['admin', 'users', 'updatePassword'],
        mutationFn: (user: UserProfile) => fetchPut(`${API_AUTH_URL}/${user.id}/password`, user),
    });
};

export const useUserCRUD: Cruddable<UserProfile> = {
    createHook: useDontCallMe,
    updateHook: useUpdateUserRole,
    deleteHook: useDontCallMe,
};
