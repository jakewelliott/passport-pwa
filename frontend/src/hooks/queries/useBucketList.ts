import { dbg } from '@/lib/debug';
import { API_BUCKET_LIST_URL, API_COMPLETED_BUCKET_LIST_ITEMS_URL, fetchGet, fetchPost } from '@/lib/fetch';
import type { BucketListCompletion, BucketListItem } from '@/types';
import { useMutation, useQuery } from '@tanstack/react-query';
import { useLocation } from '../useLocation';

/** Gets all bucket list items for all parks */
const useBucketListItems = () => {
  return useQuery({
    queryKey: ['bucketListItems'],
    queryFn: () => fetchGet(API_BUCKET_LIST_URL),
  });
};

/** Gets all completed bucket list items for a user */
const useCompletedBucketListItems = () => {
  return useQuery({
    queryKey: ['completedBucketListItems'],
    queryFn: () => fetchGet(API_COMPLETED_BUCKET_LIST_ITEMS_URL),
  });
};

/** Toggles the completion status of a bucket list item for a user
 *  Call mutate with the itemId to toggle the completion status
 */
const useToggleCompletion = () => {
  const { refetch } = useCompletedBucketListItems();
  const { geopoint } = useLocation();

  dbg('MISC', 'utc', geopoint);

  return useMutation({
    mutationFn: async (itemId: number) => {
      const response = await fetchPost(`${API_BUCKET_LIST_URL}/${itemId}`, {
        latitude: geopoint?.latitude,
        longitude: geopoint?.longitude,
        inaccuracyRadius: geopoint?.accuracy,
      });
      refetch();
      return response;
    },
  });
};

/** Call with parkId as undefined to get all bucket list items and completed items */
export const useBucketList = (parkId?: number) => {
  const { data: allItems, isLoading: isLoadingAllItems } = useBucketListItems();
  const { data: allCompletedItems, isLoading: isLoadingCompletedItems } = useCompletedBucketListItems();
  const { mutate: toggleCompletion, isPending: isMarkingAsComplete } = useToggleCompletion();

  const isLoading = isLoadingAllItems || isLoadingCompletedItems || isMarkingAsComplete;

  // filtering logic
  const parkItems = allItems?.filter((item: BucketListItem) => item.parkId === parkId);
  const parkCompletedItems = allCompletedItems?.filter((item: BucketListCompletion) => item.itemId === parkId);

  return {
    items: parkId === undefined ? allItems : parkItems,
    completed: parkId === undefined ? allCompletedItems : parkCompletedItems,
    toggleCompletion,
    isLoading,
  };
};
