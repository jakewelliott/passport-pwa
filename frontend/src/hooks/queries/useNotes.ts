import { dbg } from '@/lib/debug';
import { API_NOTES_URL, fetchGet, fetchPost } from '@/lib/fetch';
import type { ParkNote } from '@/types';
import { useMutation, useQuery } from '@tanstack/react-query';

export const useNote = (parkId: number) => {
  dbg('HOOK', 'useNote', parkId);

  return useQuery<ParkNote>({
    queryKey: ['notes', parkId],
    queryFn: () => fetchGet(`${API_NOTES_URL}/${parkId}`),
  });
};

export const useGetAllNotes = () => {
  dbg('HOOK', 'useGetAllNotes');

  return useQuery<ParkNote[]>({
    queryKey: ['notes'],
    queryFn: async () => await fetchGet(API_NOTES_URL),
  });
};

export const useUpdateNote = () => {
  const { refetch } = useGetAllNotes();
  return useMutation({
    mutationFn: ({ parkId, note }: { parkId: number; note: string }) =>
      fetchPost(`${API_NOTES_URL}/${parkId}`, { note: note, updatedAt: new Date() }),
    onSuccess: () => {
      dbg('HOOK', 'useUpdateNote', 'refetching all notes...');
      refetch();
    },
  });
};
