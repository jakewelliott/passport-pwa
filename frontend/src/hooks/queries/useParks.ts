import { useQuery } from '@tanstack/react-query';
import type { Park, ParkAbbreviation } from '@/lib/mock/types';
import { api } from '@/lib/mock/api';

// ADAM:
// This is a very simple query hook.
// It's a good starting point for most queries.
// It's a good idea to have a query hook for each endpoint.

// useQuery returns an object with some useful properties:
// - data: the data returned from the query
// - isLoading: a boolean that indicates if the query is loading (data is undefined)
// - isError: a boolean that indicates if the query has an error
// - error: the error returned from the query

export const useParks = () => {
  return useQuery<Park[]>({
    queryKey: ['parks'],
    queryFn: (): Park[] => api.getParks(),
  });
};

export const usePark = (code: ParkAbbreviation) => {
  return useQuery<Park>({
    queryKey: ['park', code],
    queryFn: (): Park => api.getPark(code),
    enabled: !!code,
  });
};
