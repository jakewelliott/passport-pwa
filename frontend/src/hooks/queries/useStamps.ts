import type { ParkAbbreviation } from '@/lib/mock/types';
import { useUser } from './useUser';
import { api } from '@/lib/mock/api';
import { useQuery } from '@tanstack/react-query';
/**
 * Gets all of the local user's stamps, empty array if user is not loaded
 * @returns The stamps for the user
 */
export const useStamps = () => {
  // re-use our query hooks whenever possible
  const { data: user } = useUser();
  return useQuery({
    queryKey: ['stamps', user?.id],
    queryFn: () => api.getUserStampsByID(user?.id || 0),
  });
};

/**
 * Gets a single stamp for a user, null if stamp not collected yet
 * @param code The code of the park to get the stamp for
 * @returns The stamp for the user
 */
export const useStamp = (code: ParkAbbreviation) => {
  // re-use our query hooks whenever possible
  const { data, isLoading } = useStamps();
  return { data: data?.find((stamp) => stamp.code === code) || null, isLoading };
};
