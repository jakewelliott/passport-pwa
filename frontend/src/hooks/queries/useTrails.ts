import { API_TRAILS_URL, fetchGet } from '@/lib/fetch';
import type { Trail } from '@/types';
import { useQuery } from '@tanstack/react-query';

export const useTrails = () => {
    return useQuery<Trail[]>({
        queryKey: ['trails'],
        queryFn: async () => await fetchGet(API_TRAILS_URL),
    });
};
