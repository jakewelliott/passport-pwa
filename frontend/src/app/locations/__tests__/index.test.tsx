import { useParks } from '@/hooks/queries/useParks';
import { setupTestEnv } from '@/lib/testing/test-wrapper';
import { beforeAll, describe, expect, it } from 'vitest';
import Locations from '../index';

const { render, checkHook } = setupTestEnv();

describe('Locations', () => {
    beforeAll(async () => {
        await checkHook(useParks, 'useParks');
    });

    it('matches snapshot', async () => {
        const { container } = render(<Locations />);
        expect(container).toMatchSnapshot();
    });

    // ADAM: add tests for filter buttons, definitely in another file tho
});
