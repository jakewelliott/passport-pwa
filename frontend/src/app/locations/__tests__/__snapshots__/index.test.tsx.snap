// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Locations > matches snapshot 1`] = `
<div>
  <div
    class="relative mx-3 mt-3 flex gap-2"
  >
    <div
      class="relative flex-1"
    >
      <input
        class="w-full rounded-lg border border-system_gray p-3 pr-7 focus:border-secondary_darkteal focus:outline-none focus:ring-1 focus:ring-secondary_darkteal focus:ring-opacity-100"
        placeholder="Search parks..."
        type="text"
        value=""
      />
    </div>
    <button
      class="w-auto rounded-lg border border-system_gray bg-system_white px-3 py-2 transition-all duration-200"
      style="display: inline;"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
        />
      </svg>
    </button>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/CABE"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Carolina Beach State Park
          </h3>
          <p>
            Carolina Beach, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/CACR"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Carvers Creek State Park
          </h3>
          <p>
            Spring Lake, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/CHRO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Chimney Rock State Park
          </h3>
          <p>
            Chimney Rock, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/CLNE"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Cliffs of the Neuse State Park
          </h3>
          <p>
            Seven Springs, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/CRMO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Crowders Mountain State Park
          </h3>
          <p>
            Kings Mountain, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/DISW"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Dismal Swamp State Park
          </h3>
          <p>
            South Mills, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/ELKN"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Elk Knob State Park
          </h3>
          <p>
            Todd, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/ENRI"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Eno River State Park
          </h3>
          <p>
            Durham, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/FALA"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Falls Lake State Recreation Area
          </h3>
          <p>
            Wake Forest, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/FOFI"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Fort Fisher State Recreation Area
          </h3>
          <p>
            Kure Beach, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/FOMA"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Fort Macon State Park
          </h3>
          <p>
            Atlantic Beach, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/GOCR"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Goose Creek State Park
          </h3>
          <p>
            Washington, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/GORG"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Gorges State Park
          </h3>
          <p>
            Sapphire, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/GRMO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Grandfather Mountain State Park
          </h3>
          <p>
            Banner Elk, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/HABE"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Hammocks Beach State Park
          </h3>
          <p>
            Swansboro, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/HARO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Hanging Rock State Park
          </h3>
          <p>
            Danbury, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/HARI"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Haw River State Park
          </h3>
          <p>
            Browns Summit, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/JORI"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Jockey's Ridge State Park
          </h3>
          <p>
            Nags Head, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/JONE"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Jones Lake State Park
          </h3>
          <p>
            Elizabethtown, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/JORD"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Jordan Lake State Recreation Area
          </h3>
          <p>
            Apex, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/KELA"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Kerr Lake State Recreation Area
          </h3>
          <p>
            Henderson, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/LAJA"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Lake James State Park
          </h3>
          <p>
            Nebo, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/LANO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Lake Norman State Park
          </h3>
          <p>
            Troutman, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/LAWA"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Lake Waccamaw State Park
          </h3>
          <p>
            Lake Waccamaw, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/LURI"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Lumber River State Park
          </h3>
          <p>
            Orrum, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/MARI"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Mayo River State Park
          </h3>
          <p>
            Mayodan, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/MEMO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Medoc Mountain State Park
          </h3>
          <p>
            Hollister, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/MEMI"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Merchants Millpond State Park
          </h3>
          <p>
            Gatesville, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/MOMO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Morrow Mountain State Park
          </h3>
          <p>
            Albemarle, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/MOJE"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Mount Jefferson State Natural Area
          </h3>
          <p>
            West Jefferson, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/MOMI"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Mount Mitchell State Park
          </h3>
          <p>
            Burnsville, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/NERI"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            New River State Park
          </h3>
          <p>
            Laurel Springs, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/OCMO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Occoneechee Mountain State Natural Area
          </h3>
          <p>
            Hillsborough, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/PETT"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Pettigrew State Park
          </h3>
          <p>
            Creswell, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/PIMO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Pilot Mountain State Park
          </h3>
          <p>
            Pinnacle, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/RARO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Raven Rock State Park
          </h3>
          <p>
            Lillington, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/REMO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Rendezvous Mountain
          </h3>
          <p>
            Purlear, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/SILA"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Singletary Lake State Park
          </h3>
          <p>
            Kelly, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/SOMO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            South Mountains State Park
          </h3>
          <p>
            Connelly Springs, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/STMO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Stone Mountain State Park
          </h3>
          <p>
            Roaring Gap, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/WEWO"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            Weymouth Woods Sandhills Nature Preserve
          </h3>
          <p>
            Southern Pines, NC
          </p>
        </div>
      </div>
    </a>
  </div>
  <div
    class="m-3"
    data-testid="park"
  >
    <a
      class="text-supporting_inactiveblue no-underline"
      data-discover="true"
      href="/locations/WIUM"
    >
      <div
        class="flex items-center bg-supporting_lightblue p-3 shadow-md"
        style="border-radius: 15px;"
      >
        <div
          class="flex flex-col gap-1"
        >
          <h3>
            William B. Umstead State Park
          </h3>
          <p>
            Raleigh, NC
          </p>
        </div>
      </div>
    </a>
  </div>
</div>
`;
